modules/zarr/src/lib/utils.ts

```diff
@@ -0,0 +1,106 @@
+import { openGroup } from 'zarr';
+import type { ZarrArray } from 'zarr';
+import type { PixelSource } from '../types';
+
+interface Multiscale {
+  datasets: { path: string }[];
+  version?: string;
+}
+
+export interface RootAttrs {
+  multiscales: Multiscale[];
+}
+
+export async function loadMultiscales(store: ZarrArray['store'], path = '') {
+  const grp = await openGroup(store, path);
+  const rootAttrs = (await grp.attrs.asObject()) as RootAttrs;
+
+  let paths = ['0'];
+  if ('multiscales' in rootAttrs) {
+    const { datasets } = rootAttrs.multiscales[0];
+    paths = datasets.map(d => d.path);
+  }
+
+  const data = paths.map(path => grp.getItem(path));
+  return {
+    data: (await Promise.all(data)) as ZarrArray[],
+    rootAttrs
+  };
+}
+
+function prevPowerOf2(x: number) {
+  return 2 ** Math.floor(Math.log2(x));
+}
+
+export function guessTileSize(arr: ZarrArray) {
+  const interleaved = isInterleaved(arr.shape);
+  const [yChunk, xChunk] = arr.chunks.slice(interleaved ? -3 : -2);
+  const size = Math.min(yChunk, xChunk);
+  // deck.gl requirement for power-of-two tile size.
+  return prevPowerOf2(size);
```

Comment:
How does this work when the inferred `tileSize` is no longer equal to the actual chunk size (when not already a power of two). Do we need any special handling for non-square chunks?

Reply:
The ZarrPixelSource handles non-square chunking. It will fetch and stitch together additional chunks if necessary to satisfy the tilesize.